### PR TITLE
Add WallMock and Unit Tests

### PR DESCRIPTION
**Pull request by pscally1005**
# Description
Add MockBukkit functionality for mocking of Walls using WallMock

# Checklist
The following items should be checked before the pull request can be merged.

- [x] Unit tests added.
- [x] Code follows existing code format.

# Info on creating a pull request
- Make sure that unit tests are added which test the relevant changes.
- Make sure that the changes follow the existing code format.

# For maintainer
When a PR is approved, the maintainer should label this PR with `release/*`.

- If the PR fixes a bug in MockBukkit, add the `release/patch` Tag.
- If the PR adds a new feature to MockBukkit, add  the `release/minor` Tag.

Note that a PR that fixes an `UnimplementedOperationException` should be considered a new version and not a bugfix.

### CODE REVIEW

File: src/main/java/be/seeseemelk/mockbukkit/block/data/WallMock.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import org.bukkit.block.data.type.Wall;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Set;
+
+import static be.seeseemelk.mockbukkit.block.data.BlockDataKey.HEIGHT;
+//import static be.seeseemelk.mockbukkit.block.data.BlockDataKey.EAST;
+//import static be.seeseemelk.mockbukkit.block.data.BlockDataKey.WEST;
+//import static be.seeseemelk.mockbukkit.block.data.BlockDataKey.NORTH;
+//import static be.seeseemelk.mockbukkit.block.data.BlockDataKey.SOUTH;
```
Comment (thelooter) [inline src/main/java/be/seeseemelk/mockbukkit/block/data/WallMock.java]: Can these be removed

**Review by thelooter — CHANGES_REQUESTED** [review]

File: src/main/java/be/seeseemelk/mockbukkit/block/data/WallMock.java (hunk 1 above)
Comment (pscally1005) [reply src/main/java/be/seeseemelk/mockbukkit/block/data/WallMock.java]: Lines have been fixed, thank you

**Review by Insprill — CHANGES_REQUESTED** [review]
I haven't done a full review, but a few things I noticed giving it a once over:

- Unrelated Javadoc/code style changes should be done in their own PR.
- There's some code style errors. Please run IntelliJ's reformatter to fix those.
- What's the height of walls? If it's related to where they're connected that will need to be updated when blocks are set in the world.

File: src/main/java/be/seeseemelk/mockbukkit/block/data/WallMock.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		setUp(false);
+		setWaterlogged(false);
+	}
+
+	/**
+	 * @return the set of possible block faces for {@link Wall}
+	 */
+	public @NotNull Set<BlockFace> getFaces()
+	{
+		return Set.of(BlockFace.EAST, BlockFace.NORTH, BlockFace.WEST, BlockFace.SOUTH);
```
Comment (TheBusyBiscuit) [inline src/main/java/be/seeseemelk/mockbukkit/block/data/WallMock.java:43]: I feel like this could/should be a `private static final` field instead.
No need to instantiate this on every method call or even for every instance of this class.

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
Looks pretty good for the most part but there is a change which I think is important.

**Comment by thelooter** [comment]
> * What's the height of walls? If it's related to where they're connected that will need to be updated when blocks are set in the world.

@pscally1005 I think this hasn't been addressed yet. Please address it

**Comment by pscally1005** [comment]
The height are none, low, and tall, and can be set for all 4 directions (east, west, north, and south).  So a wall connected by blocks in all 4 cardinal directions will have a height of low in all 4 directions

**Comment by spscally** [comment]
> The height are none, low, and tall, and can be set for all 4 directions (east, west, north, and south). So a wall connected by blocks in all 4 cardinal directions will have a height of low in all 4 directions

@pscally1005 I think I understand the question. Let me provide an example:
- if you place a wall on flat ground with nothing around it, each direction's `height` value will be `none`
- if you place a wall one block north of the original wall:
  - the original wall's `north` will have a `height` of `low`
  - the new wall's `south` will also have a `height` of `low`
- if you then place a block on top of each wall:
  - those `low` heights become `tall`

I think the question then is asking how to make sure that those height values are properly updated when blocks around the `WallMock` are updated.

File: src/main/java/be/seeseemelk/mockbukkit/block/data/BlockDataMock.java [hunk 3]
```diff
@@ -49,6 +49,95 @@ public BlockDataMock(@NotNull Material type)
...
+		{
+			return new WallSignMock(material);
+		}
+		else if (Tag.BUTTONS.isTagged(material))
+		{
+			return new SwitchMock(material);
+		}
+		return null;
+	}
+
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/block/data/BlockDataMock.java:140]: Please revert the moving of these methods.

File: src/main/java/be/seeseemelk/mockbukkit/block/data/BlockDataKey.java [hunk 4]
```diff
@@ -1,18 +1,8 @@
...
-import org.bukkit.block.data.Lightable;
-import org.bukkit.block.data.Openable;
-import org.bukkit.block.data.Powerable;
-import org.bukkit.block.data.Waterlogged;
-import org.bukkit.block.data.type.Bed;
-import org.bukkit.block.data.type.Campfire;
-import org.bukkit.block.data.type.Slab;
-import org.bukkit.block.data.type.Stairs;
+import org.bukkit.block.data.*;
+import org.bukkit.block.data.type.*;
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/block/data/BlockDataKey.java:5]: Please disable * imports. All imports should be explicit.

File: src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java:1]: This is unrelated and should be reverted.

File: src/main/java/be/seeseemelk/mockbukkit/entity/data/EntityData.java
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/data/EntityData.java:1]: This is unrelated and should be reverted.

File: src/main/java/be/seeseemelk/mockbukkit/entity/data/EntityDataRegistry.java
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/data/EntityDataRegistry.java:1]: This is unrelated and should be reverted.

**Review by Insprill — CHANGES_REQUESTED** [review]
Merge conflicts also need resolving and the height still needs to be updated when blocks in the world change. If you can't do that, the height methods can be stubbed, with the functionality implemented another time.

**Comment by Insprill** [comment]
Closing for inactivity. Feel free to reopen if the review comments are addressed.
